core/src/main/java/org/infinispan/container/DefaultDataContainer.java

```diff
@@ -100,30 +113,60 @@ protected DefaultDataContainer(int concurrencyLevel, long thresholdSize,
             throw new IllegalArgumentException("No such eviction thread policy " + strategy);
       }
 
-      Eviction eviction;
-      switch (strategy) {
-         case FIFO:
-         case UNORDERED:
-         case LRU:
-            eviction = Eviction.LRU;
+      Caffeine<K, InternalCacheEntry<K, V>> caffeine = caffeineBuilder();
+
+      switch (thresholdPolicy) {
+         case MEMORY:
+            CacheEntrySizeCalculator<K, V> calc = new CacheEntrySizeCalculator<>(new WrappedByteArraySizeCalculator<>(
+                  new MarshalledValueEntrySizeCalculator()));
+            caffeine.weigher((k, v) -> (int) calc.calculateSize(k, v)).maximumWeight(thresholdSize);
             break;
-         case LIRS:
-            eviction = Eviction.LIRS;
-            if (thresholdPolicy == EvictionType.MEMORY) {
-               throw new IllegalArgumentException("Memory based approximation eviction cannot be used with LIRS!");
-            }
+         case COUNT:
+            caffeine.maximumSize(thresholdSize);
             break;
          default:
-            throw new IllegalArgumentException("No such eviction strategy " + strategy);
-      }
-      EntrySizeCalculator<K, InternalCacheEntry<K, V>> sizeCalculator =
-            thresholdPolicy == EvictionType.MEMORY ? new CacheEntrySizeCalculator<>(
-                    new MarshalledValueEntrySizeCalculator()) : null;
+            throw new UnsupportedOperationException("Policy not supported: " + thresholdPolicy);
+      }
+      entries = new CaffeineConcurrentMap<>(applyListener(caffeine, evictionListener).build());
+   }
+
+   private Caffeine<K, InternalCacheEntry<K, V>> applyListener(Caffeine<K, InternalCacheEntry<K, V>> caffeine, DefaultEvictionListener listener) {
+      return caffeine.executor(new WithinThreadExecutor()).removalListener((k, v, c) -> {
```

Comment:
Can you use the default executor (FJP)? This is penalizing the caller, which might be small, but could be deferred to minimize response times.

If its just for unit tests, you could provide a same-thread executor only in that context.